src/models/dirac_mixture_multi_publisher_model.py

```diff
@@ -0,0 +1,666 @@
+        self,
+        observable_directions: np.ndarray,
+        frequency_histograms_on_observable_directions: np.ndarray,
+        prior_marginal_frequency_histograms: np.ndarray,
+        ncomponents: int = 1000,
+        rng: np.random.Generator = np.random.default_rng(0),
+    ):
+        """Construct an optimizer for multivariate mixed Poisson distribution.
+
+        Args:
+            observable_directions:  An <n * p> matrix where n = #training points,
+                p = #pubs.
+
+                Each training point has
+                    input = an impression vector, and
+                    output = the frequency historgam under the impression vector.
+                The i-th row of the observable_directions matrix represents the
+                impression vector of the i-th training point.
+                In reality, the impression vector could be like [1e6, 2e6],
+                which means 1e6 impressions at pub 1 and 2e6 impressions at pub 2.
+                Computing with these raw, large numbers of impresssions may
+                introduce numerical errors. As such, we standardize impression
+                vectors as "directions".  For example, suppose
+                - training point 1 has impression vector = [1e6, 2e6],
+                - training point 2 has impression vector = [1e6, 0],
+                - training point 3 has impression vector = [0, 2e6],
+                then we say that
+                - the "baseline impression vector" = [1e6, 2e6],
+                - the 3 training points have observable_directions = [
+                    [1, 1],
+                    [1, 0],
+                    [0, 1]
+                ] respectively.
+                Observable_directions is also a more compact representation
+                Compared to impression vectors.
+
+                As another example, observable_directions = [
+                    [1, 0, 0],
+                    [0, 1, 0],
+                    [0, 0, 1],
+                    [1, 1, 0],
+                    [1, 0, 1],
+                    [0, 1, 1],
+                    [1, 1, 1]
+                ] means that we observe the frequency histogram on all the subsets
+                of 3 publishers.  Rows 1-3 are single-pub observable_directions,
+                rows 4-6 indicate 2-pub unions, and rows 7 indicates 3-pub union.
+
+                As of Apr 2022, all the training points from halo_simulator will
+                have a binary observable_direction (like the above example).
+                We allow non-binary observable_direction in the future.
+            frequency_histograms_on_observable_directions:  An <n * (F + 1)> matrix
+                where n = #training points, F = maximum frequency.
+                Its i-th row is the observed frequency histogram at the i-th
+                row of observable_directions, i.e., the i-th training point.
+            prior_marginal_frequency_histograms:  A <p * (F + 1)> matrix where
```

Comment:
Consider removing this arg and including them as single pub observable directiosn